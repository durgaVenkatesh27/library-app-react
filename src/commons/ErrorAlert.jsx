import React from 'react';
import PropTypes from 'prop-types';
import Alert from '@material-ui/lab/Alert';
import '../index.css';

const ErrorAlert = ({ errorMessage, errorSeverity }) => (
  <Alert severity={errorSeverity} className="custom-alert">{errorMessage}</Alert>
);

ErrorAlert.propTypes = {
  errorMessage: PropTypes.string.isRequired
};

ErrorAlert.defaultProps = {
  errorMessage: '',
  errorSeverity: 'error'
};
export default ErrorAlert;