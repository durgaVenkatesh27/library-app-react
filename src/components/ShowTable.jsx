import React from 'react';
import * as _ from 'lodash';
import { withStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import { MaintainLibraryLocale } from '../locale';
import Button from '@material-ui/core/Button';

const StyledTableCell = withStyles((theme) => ({
  head: {
    backgroundColor: "cornflowerblue",
    color: theme.palette.common.white,
  },
  body: {
    fontSize: 14,
  },
}))(TableCell);

const StyledTableRow = withStyles((theme) => ({
  root: {
    '&:nth-of-type(odd)': {
      backgroundColor: theme.palette.action.hover,
    },
  },
}))(TableRow);

const ShowTable = props => {
  const { tableColumn, tableData, section, functionCall } = props;

  return (
    <div>
      <TableContainer component={Paper} >
        <Table aria-label="customized table" className={section}>
          <TableHead>
            <TableRow>
              {
                tableColumn.map((column, index) => (
                  <StyledTableCell
                    key={index}
                    align={"center"}
                    id={`${column}ColumnHeader`}
                  > {column}</StyledTableCell>
                ))
              }
            </TableRow>
          </TableHead>
          <TableBody>
            {_.isEmpty(tableData) ? " No data " : tableData.map((row, index) => (

              section === 'user' ? (
                <StyledTableRow key={row.id} id="tableBody">
                  <StyledTableCell align="center">{row.id}</StyledTableCell>
                  <StyledTableCell align="center">{row.name}</StyledTableCell>
                  <StyledTableCell align="center">{row.bookList}</StyledTableCell>
                </StyledTableRow>)
                : (section === 'book' ?
                  (<StyledTableRow key={row.id} id="tableBody">
                    <StyledTableCell align="center">{row.id}</StyledTableCell>
                    <StyledTableCell align="center">{row.bookTitle}</StyledTableCell>
                  </StyledTableRow>)
                  : (section === 'borrow' ? (<StyledTableRow key={row.id} id="tableBody">
                    <StyledTableCell align="center">{row.id}</StyledTableCell>
                    <StyledTableCell align="center">{row.bookTitle}</StyledTableCell>
                    <StyledTableCell align="center">
                      <Button
                        variant="contained"
                        size="small"
                        color="primary"
                        style={{ 'text-transform': 'none' }}
                        type="button"
                        className={`borrow${index} primary`}
                        onClick={(e) => functionCall(e, row.id)}
                        id={`borrow${index}`}
                      >
                        {_.get(MaintainLibraryLocale, 'borrowButtonTxt', '')}
                      </Button>
                    </StyledTableCell>
                  </StyledTableRow>)
                    : (<StyledTableRow key={row.id} id="tableBody">
                      <StyledTableCell align="center">{row.id}</StyledTableCell>
                      <StyledTableCell align="center">{row.bookTitle}</StyledTableCell>
                      <StyledTableCell align="center">
                        <Button
                          variant="contained"
                          size="small"
                          color="secondary"
                          style={{ 'text-transform': 'none' }}
                          type="button"
                          onClick={(e) => functionCall(e, row.id)}
                          id={`borrowList${index}`}
                        >
                          {_.get(MaintainLibraryLocale, 'returnButtonTxt', '')}
                        </Button>
                      </StyledTableCell>
                    </StyledTableRow>)

                  ))
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </div>
  );
}


export default ShowTable;