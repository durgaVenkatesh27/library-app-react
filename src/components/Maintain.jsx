import React, { useState } from 'react';
import * as _ from 'lodash';
import { MaintainLibraryLocale } from '../locale/index';
import ShowTable from './ShowTable';
import '../index.css';
import ErrorAlert from '../commons/ErrorAlert';
import * as api from '../commons/Api';


const Maintain = props => {
    const { userInfo, bookInfo, updateUserData, updateBookData } = props;
    const [selectedUser, setSelectedUser] = useState(-1);
    const [avlBookInfo, setAvlBookInfo] = useState({});
    const [userBookInfo, setUserBookInfo] = useState({});
    const [error, setError] = useState({ isError: false, message: '', severity: 'error' });

    React.useEffect(() => {
        const availableBooks = bookInfo.filter(book => book.isAvailable);
        setAvlBookInfo(availableBooks);

        if (selectedUser !== -1)
            updateUserBookInfo(selectedUser);
    }, [bookInfo, userInfo]);


    const handleChange = (e) => {
        const { target: { value } } = e;
        const errorObj = {
            isError: value === -1 ? true : false,
            message: value === -1 ? _.get(MaintainLibraryLocale, 'notValidToBorrowMsg', '') : '',
            severity: 'error'
        };
        setError(errorObj);
        setSelectedUser(parseInt(value));
        if (!errorObj.isError)
            updateUserBookInfo(value);
    }
    const updateUserBookInfo = (val) => {
        const user = userInfo.find(user => user.id === parseInt(val));
        if (!_.isEmpty(user.bookList)) {
            setUserBookInfo(bookInfo.filter(book => user.bookList.includes(book.id)));
        } else {
            setUserBookInfo([])
        }
    }

    const borrowBook = (e, id) => {
        const errorObj = {
            showMsg: false,
            isError: userBookInfo.length === 2 ? true : false,
            message: userBookInfo.length === 2 ? _.get(MaintainLibraryLocale, 'notValidToBorrowMsg', '') : '',
            severity: 'error'
        };
        if (selectedUser === -1) {
            errorObj.isError = true;
            errorObj.message = _.get(MaintainLibraryLocale, 'userNotAvailable', '');
            errorObj.severity = 'error';
        }
        setError(errorObj);
        if (!errorObj.isError)
            api.updateUserBook(
                `http://localhost:9090/maintain/borrowBook/${selectedUser}/${id}`,
                updateUserData,
                updateBookData,
                setError,
                _.get(MaintainLibraryLocale, 'borrowSuccessMessage', '')
            );
    }

    const returnBook = (e, id) => {
        api.updateUserBook(
            `http://localhost:9090/maintain/returnBook/${selectedUser}/${id}`,
            updateUserData,
            updateBookData,
            setError,
            _.get(MaintainLibraryLocale, 'returnSuccessMessage', '')
        );

    }
    return (
        <div>
            {
                (_.get(error, 'isError', false) || _.get(error, 'showMsg', false))
                && (<ErrorAlert errorMessage={_.get(error, 'message', '')} errorSeverity={_.get(error, 'severity', '')} />)
            }
            <form className="maintain-info">

                <label htmlFor="userNameSelect">
                    {_.get(MaintainLibraryLocale, 'selectUser', '')}
                </label>
                <select
                    className="custom-input custom-select"
                    name="userNameSelect"
                    id="userNameSelect"
                    value={selectedUser}
                    onChange={(e) => handleChange(e)}
                >
                    <option key="-1" value="-1">Select</option>
                    {
                        userInfo.map((row, index) => {
                            return (<option key={index} value={row.id}>{row.name}</option>)
                        })
                    }
                </select>
                {selectedUser !== -1 &&
                    <div>
                        <div className="borrow-list">{_.get(MaintainLibraryLocale, 'returnBookHeader', '')}</div>
                        <ShowTable
                            tableColumn={_.get(MaintainLibraryLocale, 'bookBorrowColumns', '')}
                            tableData={userBookInfo}
                            section='return'
                            functionCall={returnBook}
                        />
                    </div>}
            </form>
            <div>
                <div className="borrow-list">{_.get(MaintainLibraryLocale, 'borrowBookHeader', '')}</div>
                {_.isEmpty(avlBookInfo) ?
                    (<div >{_.get(MaintainLibraryLocale, 'noBooksAvaliable', '')}</div>)
                    : <ShowTable
                        tableColumn={_.get(MaintainLibraryLocale, 'bookBorrowColumns', '')}
                        tableData={avlBookInfo}
                        section='borrow'
                        functionCall={borrowBook}
                    />
                }
            </div>
        </div>
    );
}

export default Maintain;