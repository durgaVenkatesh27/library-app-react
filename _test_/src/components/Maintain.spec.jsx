import React from 'react';
import { mount, shallow } from 'enzyme';
import Maintain from '../../../src/components/Maintain';
import ShowTable from '../../../src/components/ShowTable';
import ErrorAlert from '../../../src/commons/ErrorAlert';


describe('Book component Tests', () => {
    let props;
    global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () => Promise.resolve({ rates: { CAD: 1.42 } }),
        })
    );
    beforeEach(() => {
        fetch.mockClear();
        props = {

            bookInfo: [{ id: 1, bookTitle: 'Sample', isAvailable: true }],
            userInfo: [{ id: 1, name: "sample", books: "1", bookList: [1] }],
            updateUserInfo: jest.fn(),
            updateBookInfo: jest.fn()
        }
    })
    it('should render Maintain Component', () => {
        const renderedModule = shallow(<Maintain {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });
    it('should render Library Table Component', () => {
        const renderedModule = mount(<Maintain {...props} />);
        expect(renderedModule).toMatchSnapshot();
        expect(renderedModule.find(ShowTable).length).toBe(1);
        expect(renderedModule.find('.borrow').length).toBe(3);
    });
    it('should render Borrowed list Table - fails', () => {
        const renderedModule = mount(<Maintain {...props} />);
        expect(renderedModule).toMatchSnapshot();
        expect(renderedModule.find('.return').length).toBe(0);
    });
    it('should render Borrowed list Table success', () => {
        const renderedModule = mount(<Maintain {...props} />);

        renderedModule.find("#userNameSelect").simulate("change", {
            target: { value: 1, selectedIndex: 1 }
        });
        expect(renderedModule).toMatchSnapshot();
        expect(renderedModule.find('.return').length).toBe(3);
    });
    it('should show error -  borrow a book when user Not selected', () => {
        Maintain.prototype.setState = jest.fn();
        const renderedModule = mount(<Maintain {...props} />);
        expect(renderedModule).toMatchSnapshot();
        renderedModule.find("button#borrow0").simulate("click");
        expect(renderedModule.find(ErrorAlert).length).toBe(1);
    });

    it('should borrow a book success', () => {

        let user = [];
        const updateUserData = jest.fn(() => {
            user = [{ id: 1, name: 'Sample', books: "1", bookList: [1] }];
            return true;
        });
        const updateBookData = jest.fn(() => {
            user = [{ id: 1, bookTitle: 'Sample', isAvailable: false }];
            return true;
        })
        Maintain.prototype.setState = jest.fn();

        const renderedModule = mount(<Maintain {...props} updateUserData={updateUserData} updateBookData={updateBookData} />);
        renderedModule.find("#userNameSelect").simulate("change", {
            target: { value: 1, selectedIndex: 1 }
        });
        expect(renderedModule).toMatchSnapshot();
        renderedModule.find("button#borrow0").simulate("click");
        setTimeout(function () { expect(renderedModule.find("button#borrow0").length).toBe(0); }, 3000);
    });

    it('should show error - borrow a book when user cross his limit', () => {
        Maintain.prototype.setState = jest.fn();
        props.userInfo = [{ id: 1, name: "sample", books: "1,2", bookList: ["1", "2"] }]
        const renderedModule = mount(<Maintain {...props} />);
        expect(renderedModule).toMatchSnapshot();
        renderedModule.find("button#borrow0").simulate("click");
        expect(renderedModule.find(ErrorAlert).length).toBe(1);
    });


    it('should return a book success', () => {

        props.userInfo = [{ id: 1, name: "sample", books: "1,2", bookList: ["1", "2"] }];
        let user = [];
        const updateUserData = jest.fn(() => {
            user = [{ id: 1, name: 'Sample', books: "1", bookList: [1] }];
            return true;
        });
        const updateBookData = jest.fn(() => {
            user = [{ id: 1, bookTitle: 'Sample', isAvailable: false }];
            return true;
        })
        Maintain.prototype.setState = jest.fn();

        const renderedModule = mount(<Maintain {...props} updateUserData={updateUserData} updateBookData={updateBookData} />);
        renderedModule.find("#userNameSelect").simulate("change", {
            target: { value: 1, selectedIndex: 1 }
        });


        setTimeout(function () {
            renderedModule.find("button#borrowList0").simulate("click");
            expect(renderedModule.find("button#borrowList0").length).toBe(0);
        }, 3000);
    });
});