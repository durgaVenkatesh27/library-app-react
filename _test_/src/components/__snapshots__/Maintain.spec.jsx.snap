// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book component Tests should borrow a book success 1`] = `ReactWrapper {}`;

exports[`Book component Tests should render Borrowed list Table - fails 1`] = `ReactWrapper {}`;

exports[`Book component Tests should render Borrowed list Table success 1`] = `ReactWrapper {}`;

exports[`Book component Tests should render Library Table Component 1`] = `ReactWrapper {}`;

exports[`Book component Tests should render Maintain Component 1`] = `ShallowWrapper {}`;

exports[`Book component Tests should show error -  borrow a book when user Not selected 1`] = `ReactWrapper {}`;

exports[`Book component Tests should show error - borrow a book when user cross his limit 1`] = `ReactWrapper {}`;
